app/custom_auth.py

```diff
@@ -0,0 +1,89 @@
+from django.contrib import auth
+from django.conf import settings
+from django.contrib.auth.models import User, Group
+from django.core.exceptions import PermissionDenied
+
+import logging
+
+logger = logging.getLogger(__name__)
+
+class CustomSSOMiddleware: # pragma: no cover
+    """
+    A custom middleware that allows SSO via HTTP headers passed in the request
+    
+    Example 
+    -------
+    HTTP_X_AUTH_PROXY_USER : User 
+    HTTP_X_AUTH_PROXY_GROUPS: Admin,Editor
+
+    Summary
+    -------
+    (1) If HTTP header HTTP_X_AUTH_PROXY_USER with the user name is not passed no further processing is done by the middleware.
+    (2) If HTTP header with the user name is passed then:
+        (2.1) Authenciate the user if already exist in the database. 
+        (2.2) Create user if does not exist in the database and then autenicate the user. 
+    (3) If IS_SUPER flag is set to True in the settings.py then ingnore HTTP_X_AUTH_PROXY_GROUPS and grant the user Admin level privileges
+    (4) If HTTP_X_AUTH_PROXY_GROUPS header is present and contains Admin group then grant the user Admin level privileges. 
+    (5) If HTTP_X_AUTH_PROXY_GROUPS header is absent or does not contain the Admin group then revoke user's Admin level privleges.  
+    (6) User is now signed in and session cookies are set.  
+    ...
+    """
+    def __init__(self, get_response):
+        self.get_response = get_response
+     
+    def __call__(self, request):  
+        if not request.user.is_authenticated:
+            try:
+                username = request.META[settings.HEADER_AUTH_PROXY_USER]
+            except KeyError:
+                logger.exception(f"User name header {settings.HEADER_AUTH_PROXY_USER} not found. Continue without user authenication." )
```

Comment:
This will log an exception for _every_ deployment of doccano modulo the few ones that use the feature... really not using the feature is the expected behavior so we should only log at debug level here.